counter/src/main/java/org/infinispan/counter/impl/CounterModuleLifecycle.java

```diff
@@ -0,0 +1,361 @@
+import org.infinispan.configuration.cache.CacheMode;
+import org.infinispan.configuration.cache.Configuration;
+import org.infinispan.configuration.cache.ConfigurationBuilder;
+import org.infinispan.configuration.global.GlobalConfiguration;
+import org.infinispan.context.Flag;
+import org.infinispan.counter.api.CounterConfiguration;
+import org.infinispan.counter.api.CounterManager;
+import org.infinispan.counter.api.CounterState;
+import org.infinispan.counter.api.CounterType;
+import org.infinispan.counter.api.Storage;
+import org.infinispan.counter.api.StrongCounter;
+import org.infinispan.counter.api.WeakCounter;
+import org.infinispan.counter.configuration.AbstractCounterConfiguration;
+import org.infinispan.counter.configuration.CounterManagerConfiguration;
+import org.infinispan.counter.configuration.CounterManagerConfigurationBuilder;
+import org.infinispan.counter.configuration.Reliability;
+import org.infinispan.counter.configuration.StrongCounterConfiguration;
+import org.infinispan.counter.configuration.WeakCounterConfiguration;
+import org.infinispan.counter.impl.entries.CounterKey;
+import org.infinispan.counter.impl.entries.CounterValue;
+import org.infinispan.counter.impl.function.AddFunction;
+import org.infinispan.counter.impl.function.CompareAndSetFunction;
+import org.infinispan.counter.impl.function.InitializeCounterFunction;
+import org.infinispan.counter.impl.function.ReadFunction;
+import org.infinispan.counter.impl.function.ResetFunction;
+import org.infinispan.counter.impl.interceptor.CounterConfigurationInterceptor;
+import org.infinispan.counter.impl.interceptor.CounterInterceptor;
+import org.infinispan.counter.impl.listener.CounterFilterAndConvert;
+import org.infinispan.counter.impl.metadata.ConfigurationMetadata;
+import org.infinispan.counter.impl.strong.BoundedStrongCounter;
+import org.infinispan.counter.impl.strong.StrongCounterKey;
+import org.infinispan.counter.impl.strong.UnboundedStrongCounter;
+import org.infinispan.counter.impl.weak.WeakCounterImpl;
+import org.infinispan.counter.impl.weak.WeakCounterKey;
+import org.infinispan.counter.logging.Log;
+import org.infinispan.counter.util.Utils;
+import org.infinispan.factories.GlobalComponentRegistry;
+import org.infinispan.interceptors.impl.EntryWrappingInterceptor;
+import org.infinispan.lifecycle.AbstractModuleLifecycle;
+import org.infinispan.lifecycle.ModuleLifecycle;
+import org.infinispan.manager.EmbeddedCacheManager;
+import org.infinispan.registry.InternalCacheRegistry;
+import org.infinispan.transaction.TransactionMode;
+import org.kohsuke.MetaInfServices;
+
+/**
+ * It register a {@link CounterManagerImpl} to each {@link EmbeddedCacheManager} started and starts the cache on it.
+ *
+ * @author Pedro Ruivo
+ * @since 9.0
+ */
+@MetaInfServices(value = ModuleLifecycle.class)
+public class CounterModuleLifecycle extends AbstractModuleLifecycle {
+
+   public static final String COUNTER_CACHE_NAME = "___counters";
```

Comment:
yes all counters share the `num-onwers` and `availability`. 
About the `num-owner` I don't think a user would have different `num-owners` per counter if the counters are safe (in terms of failures).
About the `availability`, lets see how the users use it. If needed, I'll move it to be configure per counter. But, to be honest, I prefer not to create a bunch of caches for the counters if we try to support any `num-owner`/`availability` configuration...